tool/berty-client/client.go

```diff
@@ -0,0 +1,320 @@
+package main
+
+import (
+	"context"
+	"encoding/base64"
+	"flag"
+	"fmt"
+	"log"
+	"os"
+	"time"
+
+	"berty.tech/berty/v2/go/pkg/bertylinks"
+	"berty.tech/berty/v2/go/pkg/messengertypes"
+	"berty.tech/berty/v2/go/pkg/protocoltypes"
+
+	grpc_zap "github.com/grpc-ecosystem/go-grpc-middleware/logging/zap"
+	"github.com/oklog/run"
+	"github.com/peterbourgon/ff/v3/ffcli"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+)
+
+// flags
+var (
+	rootFlagSet = flag.NewFlagSet("client", flag.ExitOnError)
+
+	hostaddr = rootFlagSet.String("host", "127.0.0.1:9091", "berty daemon addr")
+	verbose  = rootFlagSet.Bool("v", false, "increase log verbosity")
+
+	groupFlagSet = flag.NewFlagSet("group", flag.ExitOnError)
+	groupPK      = groupFlagSet.String("pk", "", "group pk (base64)")
+	groupName    = groupFlagSet.String("name", "some group", "group name")
+	groupLink    = groupFlagSet.String("link", "", "group link")
+
+	contactFlagSet = flag.NewFlagSet("contact", flag.ExitOnError)
+	contactLink    = contactFlagSet.String("link", "", "shareable url")
+)
+
+type client struct {
+	logger *zap.Logger
+
+	messenger messengertypes.MessengerServiceClient
+	protocol  protocoltypes.ProtocolServiceClient
+}
+
+func (c *client) contactShare(ctx context.Context) error {
+	res, err := c.messenger.InstanceShareableBertyID(ctx, &messengertypes.InstanceShareableBertyID_Request{})
+	if err != nil {
+		return err
+	}
+
+	fmt.Println(res.GetWebURL())
+
+	return nil
+}
+
+func (c *client) contactRequest(ctx context.Context, weburl string) error {
+	_, err := c.messenger.ContactRequest(ctx, &messengertypes.ContactRequest_Request{
+		Link: weburl,
+	})
+
+	return err
+}
+
+// @TODO(gfanton): contact accept all
+
+func (c *client) groupNew(ctx context.Context) error {
+	res, err := c.protocol.MultiMemberGroupCreate(ctx, &protocoltypes.MultiMemberGroupCreate_Request{})
+	if err != nil {
+		return err
+	}
+
+	gpk := base64.RawStdEncoding.EncodeToString(res.GroupPK)
+	fmt.Println(gpk)
+
+	return nil
+}
+
+func (c *client) groupShare(ctx context.Context, name, base64pk string) error {
+	gpk, err := base64.RawStdEncoding.DecodeString(base64pk)
+	if err != nil {
+		return fmt.Errorf("unable to decode b64 pk: %w", err)
+	}
+
+	res, err := c.messenger.ShareableBertyGroup(ctx, &messengertypes.ShareableBertyGroup_Request{
+		GroupPK:   gpk,
+		GroupName: name,
+	})
+	if err != nil {
+		return err
+	}
+
+	fmt.Println(res.GetWebURL())
+
+	return nil
+}
+
+func (c *client) groupJoin(ctx context.Context, url string) error {
+	link, err := bertylinks.UnmarshalLink(url, nil) // FIXME: support passing an optional passphrase to decrypt the link
+	if err != nil {
+		return fmt.Errorf("unalbe to unmarshal link: %w", err)
+	}
+	if !link.IsGroup() {
+		return fmt.Errorf("provided link is not a group")
+	}
+
+	_, err = c.protocol.MultiMemberGroupJoin(ctx, &protocoltypes.MultiMemberGroupJoin_Request{
```

Comment:
Why not use the messenger api there too?

Reply:
I mostly taked the code from `berty mini`, I will check messenger api